File: apps/pos_app/lib/data/models/loyalty/loyalty_program_entity.dart
import 'package:floor/floor.dart';

@Entity(
  tableName: 'loyalty_programs',
  indices: [
    Index(value: ['tenant_id']),
    Index(value: ['status']),
    Index(value: ['program_type']),
  ],
)
class LoyaltyProgramEntity {
  @primaryKey
  final String id;

  @ColumnInfo(name: 'tenant_id')
  final String tenantId;

  final String name;

  @ColumnInfo(name: 'program_type')
  final String programType;

  final String status;

  @ColumnInfo(name: 'starts_at')
  final int? startsAt;

  @ColumnInfo(name: 'ends_at')
  final int? endsAt;

  @ColumnInfo(name: 'earning_rule_json')
  final String earningRuleJson;

  @ColumnInfo(name: 'eligibility_rule_json')
  final String eligibilityRuleJson;

  @ColumnInfo(name: 'config_version')
  final int configVersion;

  @ColumnInfo(name: 'created_at')
  final int createdAt;

  @ColumnInfo(name: 'updated_at')
  final int updatedAt;

  LoyaltyProgramEntity({
    required this.id,
    required this.tenantId,
    required this.name,
    required this.programType,
    required this.status,
    this.startsAt,
    this.endsAt,
    required this.earningRuleJson,
    required this.eligibilityRuleJson,
    required this.configVersion,
    required this.createdAt,
    required this.updatedAt,
  });
}

File: apps/admin_backend/src/modules/loyalty/entities/loyalty-program.entity.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  JoinColumn,
  Index,
  OneToMany,
} from 'typeorm';
import { Tenant } from '../../tenant/entities/tenant.entity';
import { RewardDefinition } from './reward-definition.entity';

export enum LoyaltyProgramType {
  SPEND_POINTS = 'SPEND_POINTS',
  PRODUCT_STAMPS = 'PRODUCT_STAMPS',
  VISIT_STAMPS = 'VISIT_STAMPS',
}

export enum LoyaltyProgramStatus {
  DRAFT = 'DRAFT',
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
}

@Entity('loyalty_programs')
@Index('idx_loyalty_programs_tenant_status', ['tenant_id', 'status'])
export class LoyaltyProgram {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  tenant_id: string;

  @ManyToOne(() => Tenant)
  @JoinColumn({ name: 'tenant_id' })
  tenant: Tenant;

  @Column({ type: 'varchar', length: 160 })
  name: string;

  @Column({
    type: 'enum',
    enum: LoyaltyProgramType,
  })
  program_type: LoyaltyProgramType;

  @Column({
    type: 'enum',
    enum: LoyaltyProgramStatus,
    default: LoyaltyProgramStatus.DRAFT,
  })
  status: LoyaltyProgramStatus;

  @Column({ type: 'timestamptz', nullable: true })
  starts_at?: Date | null;

  @Column({ type: 'timestamptz', nullable: true })
  ends_at?: Date | null;

  @Column({ type: 'jsonb' })
  earning_rule: Record<string, unknown>;

  @Column({ type: 'jsonb' })
  eligibility_rule: Record<string, unknown>;

  @Column({ type: 'int', default: 1 })
  config_version: number;

  @OneToMany(() => RewardDefinition, (reward) => reward.loyalty_program)
  rewards: RewardDefinition[];

  @CreateDateColumn({ name: 'created_at' })
  created_at: Date;

  @UpdateDateColumn({ name: 'updated_at' })
  updated_at: Date;
}
